sql: avoid string to byte conversion copies on insert path

cucaroach opened the pull request:
Originated from: https://github.com/cockroachdb/cockroach/pull/91930

Incorporate those changes and add a benchmark and add some more changes based on benchmark results.

Broken into several commits:

### json: avoid copying json.AsText where possible
    
Epic: None
Release note: None


### pgwire: avoid some string to byte copies

Most parse routines don't retain the input string pointer so we can
use a pointer to the input bytes in those cases.

Release note: None
Epic: None

### builtins: avoid some []byte to string copies in decode

Release note: None
Epic: None



### valueside: remove string to []byte copies

Release note: None
Epic: None


### colenc: use new UnsafeConvertBytesToString

Cosmetic change to use the new UnsafeConvertBytesToString.

Release note: None
Epic: None



### opt: avoid allocation in DBytes interning

DBytes is a string under the covers but we usually operate on them
with []byte APIs, avoid copies in these cases.

Release note: None
Epic: None


### tree: allow access to DBytes/DString/DEncodedKey raw bytes

Epic: None
Release note: None


### encoding: unsafeString -> UnsafeConvertBytesToString

Make public for use elsewhere.

Release note: None
Epic: None

### lex: avoid some string to byte slice copies

Facilitate some copy avoidance by using []byte instead of string.
Only copy when necessary in some cases.

Release note: None
Epic: None

### bench: add a large insert benchmark

This benchmark shows the reducation in allocations by the copy avoidance
changes.  Together the changes result in:

```
name                                     old time/op    new time/op    delta
SQL/Cockroach/InsertLarge/count=1000-10    18.7ms ± 4%    19.1ms ±26%     ~     (p=0.780 n=9+10)

name                                     old alloc/op   new alloc/op   delta
SQL/Cockroach/InsertLarge/count=1000-10    49.7MB ± 4%    39.7MB ±10%  -20.19%  (p=0.000 n=9+9)

name                                     old allocs/op  new allocs/op  delta
SQL/Cockroach/InsertLarge/count=1000-10     39.0k ± 5%     29.7k ± 8%  -23.83%  (p=0.000 n=10+9)
```

Release note: None
Epic: None

cucaroach committed:
tree: allow access to DBytes/DString/DEncodedKey raw bytes

Epic: None
Release note: None

cucaroach committed:
opt: avoid allocation in DBytes interning

DBytes is a string under the covers but we usually operate on them
with []byte APIs, avoid copies in these cases.

Release note: None
Epic: None

cucaroach committed:
encoding: unsafeString -> UnsafeConvertBytesToString

Make public for use elsewhere.

Release note: None
Epic: None

cucaroach committed:
colenc: use new UnsafeConvertBytesToString

Cosmetic change to use the new UnsafeConvertBytesToString.

Release note: None
Epic: None

cucaroach committed:
valueside: remove string to []byte copies

Release note: None
Epic: None

cucaroach committed:
lex: avoid some string to byte slice copies

Facilitate some copy avoidance by using []byte instead of string.
Only copy when necessary in some cases.

Release note: None
Epic: None

cucaroach committed:
builtins: avoid some []byte to string copies in decode

Release note: None
Epic: None

cucaroach committed:
pgwire: avoid some string to byte copies

Most parse routines don't retain the input string pointer so we can
use a pointer to the input bytes in those cases.

Release note: None
Epic: None

cucaroach committed:
bench: add a large insert benchmark

This benchmark shows the reducation in allocations by the copy avoidance
changes.  Together the changes result in:

```
name                                     old time/op    new time/op    delta
SQL/Cockroach/InsertLarge/count=1000-10    18.7ms ± 4%    19.1ms ±26%     ~     (p=0.780 n=9+10)

name                                     old alloc/op   new alloc/op   delta
SQL/Cockroach/InsertLarge/count=1000-10    49.7MB ± 4%    39.7MB ±10%  -20.19%  (p=0.000 n=9+9)

name                                     old allocs/op  new allocs/op  delta
SQL/Cockroach/InsertLarge/count=1000-10     39.0k ± 5%     29.7k ± 8%  -23.83%  (p=0.000 n=10+9)
```

Release note: None
Epic: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/101693)
<!-- Reviewable:end -->

msirek reviewed (approved):
Nice reduction in allocations! Maybe this should be backported to 23.1.
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 1 files at r1, 2 of 2 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, 3 of 3 files at r6, 1 of 1 files at r7, 1 of 1 files at r8, 2 of 2 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101693#-:-NTGqcowCLfmmxwP1Eeh:b-39fy8g)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rafiss)



<!-- Sent from Reviewable.io -->

cucaroach commented:
Thanks @msirek!   I'm gonna wait for @rafiss review as well for the pgwire and lex bits.

otan commented on pkg/sql/sem/builtins/builtins.go:
```diff
@@ -1164,11 +1164,11 @@ var regularBuiltins = map[string]builtinDefinition{
 					return nil, err
 				}
-				return tree.NewDBytes(tree.DBytes(res)), nil
+				return tree.NewDBytes(tree.DBytes(encoding.UnsafeConvertBytesToString(res))), nil
```
hmm, does this get re-used if someone uses another builtin that modifies bytes directly?

otan reviewed (commented):
admittedly not well versed in the unsafe Go string/bytes stuff, but the changes look good to me except maybe the builtins

cucaroach committed:
sql: Avoid copy of file upload data

All we do here is write the bytes to a stream so we don't need a copy.

Release note: None
Epic: None

cucaroach commented:
Adding @yuzefovich to double check my math on DBytes safety, my concern is that maybe the vectorized engine constructs DBytes directly from coldata.Bytes memory.

cucaroach committed:
builtins: avoid multiple data copies in convert_from

We copy the DBytes into a []byte and then copy the []byte into a string.
Now we get a reference to the string and return that if appropriate or
iterator over string bytes directly if not.  This is all predicated
on DBytes being immutable.

Release note: None
Epic: None

yuzefovich reviewed (commented):
Nice reduction!

> to double check my math on DBytes safety, my concern is that maybe the vectorized engine constructs DBytes directly from coldata.Bytes memory.

I don't think we have any issues there. We currently always make a copy when going from vectorized `coldata.Bytes` to `DBytes` datums. For example, over in `vec_to_datum_gen.go` we make a copy by converting `[]byte` (that we get from the `Bytes` vector) to `DBytes` (which is a `string`).

> Reviewed 1 of 1 files at r1, 2 of 2 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, 3 of 3 files at r6, 1 of 1 files at r7, 1 of 1 files at r8, 2 of 2 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101693#-:-NTvE3sH8K_H-Jvtg0B1:b-wn63qs)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @cucaroach, @nvanbenschoten, @otan, and @rafiss)

___
*[`-- commits` line 54 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/101693#-NTvCodE0Jjxz7DCnJiW):*
nit: there is another place where we can do the same - in two spots in `execgen/overloads_bin.go` after which you'll need to run `./dev gen execgen` to regenerate the code. Or perhaps leave a TODO for that.
___
*[`pkg/sql/sem/builtins/builtins.go` line 1171 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/101693#gh-1175175796:-NTvDScB8PbP8xrYA5aP:bbcbsu3) ([raw file](https://github.com/cockroachdb/cockroach/blob/d463da3957254cbcd3a0f9d57f1c327c1a51948b/pkg/sql/sem/builtins/builtins.go#L1171)):*
<details><summary><i>Previously, cucaroach (Tommy Reilly) wrote…</i></summary><blockquote>

I'm DBytes are immutable based on comments here: 

https://github.com/cucaroach/cockroach/blob/d463da3957254cbcd3a0f9d57f1c327c1a51948b/pkg/sql/sem/tree/datum.go#L1503

Most mutating consumers of DBytes do this:

```
	infoBytes := []byte(tree.MustBeDBytes(row[1]))
```

Which will make a copy. 
</blockquote></details>

Yeah, this should be safe.
___
*[`pkg/sql/sem/tree/datum.go` line 381 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/101693#-NTvC1uU09lujfKd7HEd:-NTvC1uU09lujfKd7HEe:b-4e3az) ([raw file](https://github.com/cockroachdb/cockroach/blob/266bc034156c5f8963501319ebb324c950628ff0/pkg/sql/sem/tree/datum.go#L381)):*
> ```Go
> 		return nil, MakeParseError(s, types.Bytes, err)
> 	}
> 	return NewDBytes(DBytes(encoding.UnsafeConvertBytesToString(res))), nil
> ```

This function is defined in a later commit, so the build should fail on the third commit. This should be fixed. (Broken builds are bad because they make bisecting more annoying.)
___
*[`pkg/bench/bench_test.go` line 308 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101693#-NTv9BxJ5CCzapNa4dNM:-NTv9BxJ5CCzapNa4dNN:b88zbuz) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd3325d7a6208c3d04efb150a352e5a083bbcab6/pkg/bench/bench_test.go#L308)):*
> ```Go
> }
> 
> // runBenchmarkInsert benchmarks inserting count rows into a table.
> ```

nit: `s/runBenchmarkInsert/runBenchmarkInsertLarge/`, also mention how it differs from `runBenchmarkInsert`.


<!-- Sent from Reviewable.io -->

cucaroach committed:
json: avoid copying json.AsText where possible

Epic: None
Release note: None

cucaroach reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101693)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @msirek, @nvanbenschoten, @otan, @rafiss, and @yuzefovich)

___
*[`pkg/sql/sem/tree/datum.go` line 381 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/101693#-NTvC1uU09lujfKd7HEd:-NUXWXBCD5LH4PQGbzIt:b-886gsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/266bc034156c5f8963501319ebb324c950628ff0/pkg/sql/sem/tree/datum.go#L381)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

This function is defined in a later commit, so the build should fail on the third commit. This should be fixed. (Broken builds are bad because they make bisecting more annoying.)
</blockquote></details>

Fixed
___
*[`pkg/bench/bench_test.go` line 308 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101693#-NTv9BxJ5CCzapNa4dNM:-NUXXB988eAnJ_Tu9BCM:b-8fx2u7) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd3325d7a6208c3d04efb150a352e5a083bbcab6/pkg/bench/bench_test.go#L308)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: `s/runBenchmarkInsert/runBenchmarkInsertLarge/`, also mention how it differs from `runBenchmarkInsert`.
</blockquote></details>

:+1: 


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
the changes looked safe to me!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101693#-:-NWjyD7u8UdCK8PmnNMb:b-c1i4ca)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @msirek, @nvanbenschoten, @otan, and @yuzefovich)



<!-- Sent from Reviewable.io -->

cucaroach commented:
TFTRs!
bors r+
